### PR TITLE
convert websocket acceptor to have separate middleware

### PR DESCRIPTION
**Pull request by kevo1ution**
## Changes proposed in this PR

- adds two middleware for websocket upgrade request acceptor: https-redirect and maintenance-mode

## Why are we making these changes?

- websockets upgrade requests need to go through similar logic as regular requests. Both already support authentication, but not https-redirect or maintenance-mode logic. There is still some more work to be done, but this PR should make it easier to add middleware to websocket upgrade requests.

### CODE REVIEW

File: waiter/src/waiter/core.clj [hunk 1]
```diff
@@ -970,9 +970,9 @@
                                                         (request-handler request)
                                                         :else
                                                         (auth-handler request)))
-                                             jwt-authenticator (jwt/wrap-auth-handler jwt-authenticator)
-                                             oidc-authenticator (oidc/wrap-auth-handler oidc-authenticator)
-                                             true (auth/wrap-auth-cookie-handler password)))))
+                                                    jwt-authenticator (jwt/wrap-auth-handler jwt-authenticator)
+                                                    oidc-authenticator (oidc/wrap-auth-handler oidc-authenticator)
+                                                    true (auth/wrap-auth-cookie-handler password)))))
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj:975]: Please avoid unrelated indentation changes.

File: waiter/src/waiter/core.clj [hunk 2]
```diff
@@ -1196,20 +1179,38 @@
                                                   :method (some-> request .getMethod str/lower-case)
                                                   :protocol-version (.getProtocolVersion request)
                                                   :sub-protocols (some-> request .getSubProtocols seq)
-                                                  :uri (some-> request .getRequestURI .getPath)})
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj:1199]: This version is `nil`-safe

File: waiter/src/waiter/core.clj [hunk 3]
```diff
@@ -1178,16 +1147,30 @@
    :websocket-request-auth-cookie-attacher (pc/fnk [[:state passwords router-id]]
                                              (fn websocket-request-auth-cookie-attacher [request]
                                                (ws/inter-router-request-middleware router-id (first passwords) request)))
-   :websocket-request-acceptor (pc/fnk [[:state passwords server-name]
-                                        discover-service-parameters-fn]
+   :websocket-request-acceptor (pc/fnk [[:state server-name]
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj:1150]: Should this move to handler instead?

File: waiter/src/waiter/handler.clj [hunk 4]
```diff
@@ -47,7 +48,37 @@
...
+          http-request? (= :http (utils/request->scheme request))
+          https-redirect? (get-in request [:waiter-discovery :token-metadata "https-redirect"])]
+      (if (and http-request? https-redirect?)
+        (do
+          (log/info "triggering ssl redirect")
+          (-> (ssl/ssl-redirect-response request {})
+              (utils/attach-waiter-source)))
+        (handler request)))))
+
+(defn wrap-https-redirect-acceptor
```
Comment (shamsimam) [inline waiter/src/waiter/handler.clj:69]: Maybe rename to `wrap-wss-redirect-acceptor`

File: waiter/integration/waiter/websocket_integration_test.clj [hunk 5]
```diff
@@ -32,18 +32,24 @@
            (org.eclipse.jetty.websocket.api UpgradeException UpgradeRequest)
            (org.eclipse.jetty.websocket.client WebSocketClient)
            (qbits.jet.client.websocket Connection)
-           (qbits.jet.websocket WebSocket)))
+           (qbits.jet.websocket WebSocket)
+           (org.eclipse.jetty.client HttpClient)))
```
Comment (shamsimam) [inline waiter/integration/waiter/websocket_integration_test.clj:36]: lexicographic ordering

File: waiter/integration/waiter/websocket_integration_test.clj [hunk 6]
```diff
@@ -192,6 +198,152 @@
...
+                                     (swap! ws-response-atom conj (async/<! in))
+                                     (swap! ws-response-atom conj (async/<! in))
+                                     (deliver response-promise :done)
+                                     (async/close! out)))
+                                 {:middleware (fn [_ ^UpgradeRequest request]
+                                                (websocket/add-headers-to-upgrade-request! request waiter-headers))})
+                    [close-code error] (connection->ctrl-data connection)]
+                (is (= :qbits.jet.websocket/error close-code))
+                (is (instance? UpgradeException error))
+                (is (str/includes? (.getMessage error) "Unexpected HTTP Response Status Code: 503 Service Unavailable")))
```
Comment (shamsimam) [inline waiter/integration/waiter/websocket_integration_test.clj:235]: Does `custom maintenance message` show up in the response here?

File: waiter/integration/waiter/websocket_integration_test.clj [hunk 7]
```diff
@@ -192,6 +198,152 @@
...
+                                     (async/close! out)))
+                                 {:middleware (fn [_ ^UpgradeRequest request]
+                                                (.setHeader request "x-waiter-maintenance" "some value")
+                                                (websocket/add-headers-to-upgrade-request! request waiter-headers))})
+                    [close-code error] (connection->ctrl-data connection)]
+                (is (= :qbits.jet.websocket/error close-code))
+                (is (instance? UpgradeException error))
+                (is (str/includes? (.getMessage error) "Unexpected HTTP Response Status Code: 400 Bad Request")))
+              (finally
+                (delete-service waiter-url waiter-headers))))
```
Comment (shamsimam) [inline waiter/integration/waiter/websocket_integration_test.clj:265]: No service should have been started, what are we deleting here?

File: waiter/integration/waiter/websocket_integration_test.clj [hunk 8]
```diff
@@ -192,6 +198,152 @@
...
+                                     (swap! ws-response-atom conj (async/<! in))
+                                     (deliver response-promise :done)
+                                     (async/close! out)))
+                                 {:middleware (fn [_ ^UpgradeRequest request]
+                                                (.setHeader request "x-waiter-maintenance" "some value")
+                                                (websocket/add-headers-to-upgrade-request! request waiter-headers))})
+                    [close-code error] (connection->ctrl-data connection)]
+                (is (= :qbits.jet.websocket/error close-code))
+                (is (instance? UpgradeException error))
+                (is (str/includes? (.getMessage error) "Unexpected HTTP Response Status Code: 400 Bad Request")))
```
Comment (shamsimam) [inline waiter/integration/waiter/websocket_integration_test.clj:263]: Nitpick: We can create a helper macro for asserting upgrade exceptions.

File: waiter/integration/waiter/websocket_integration_test.clj [hunk 9]
```diff
@@ -192,6 +198,152 @@
...
+      (when (some-> waiter-url waiter-settings :server-options :ssl-port)
+        (testing "should upgrade if wss and https-redirect is enabled"
+          (try
+            (let [token-response (post-token waiter-url token-description)]
+              (assert-response-status token-response http-200-ok)
+              (try
+                (let [response-promise (promise)
+                      connection (ws-client/connect!
+                                   (websocket-client-factory)
+                                   (wss-url waiter-url "/websocket-auth")
```
Comment (shamsimam) [inline waiter/integration/waiter/websocket_integration_test.clj:292]: Doesn't the url need to specify the SSL port?

File: waiter/src/waiter/auth/authentication.clj [hunk 10]
```diff
@@ -379,16 +380,36 @@
...
+  authentication parameter. If successful and authentication is disabled for the token,
+  then the :skip-authentication is set to true for the request object before getting passed
+  to the handler"
+  [handler on-error]
   (fn auth-bypass-handler [{:keys [waiter-discovery] :as request}]
     (process-authentication-parameter
       waiter-discovery
-      (fn [status message]
-        (utils/clj->json-response {:error message} :status status))
+      (partial on-error request)
```
Comment (shamsimam) [inline waiter/src/waiter/auth/authentication.clj:392]: Avoid using partial function. Previous experimentation has found them to have a performance penalty. We avoid including them in potential hot paths. Please replace with a `fn` that lists out the parameters.

File: waiter/src/waiter/core.clj (hunk 3 above)
Comment (shamsimam) [reply waiter/src/waiter/core.clj:1150]: I meant `:websocket-request-acceptor` moving to handler. 
We can also choose `websocket/make-websocket-request-acceptor` as the target for the function body to aggregate all the websocket methods in one file, `websocket.clj`.

File: waiter/src/waiter/core.clj [hunk 11]
```diff
@@ -1147,38 +1147,26 @@
...
+                                 ; If adding new middleware for websocket upgrade requests, consider adding the same middleware to
+                                 ; process-request-wrapper-fn
+                                 (let [handler (-> #(ws/request-subprotocol-acceptor (:upgrade-request %) (:upgrade-response %))
+                                                   wrap-secure-request-acceptor-fn
+                                                   auth/wrap-auth-bypass-acceptor
+                                                   pr/wrap-maintenance-mode-acceptor
+                                                   handler/wrap-wss-redirect
+                                                   wrap-service-discovery-fn)]
+                                   (handler/make-websocket-request-acceptor server-name handler)))
+   :wrap-secure-request-acceptor-fn (pc/fnk [[:state passwords]]
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj:1161]: Maybe rename to `websocket-secure-request-acceptor-fn`

File: waiter/src/waiter/handler.clj [hunk 12]
```diff
@@ -48,23 +48,44 @@
             [waiter.util.ring-utils :as ru]
             [waiter.util.semaphore :as semaphore]
             [waiter.util.utils :as utils])
-  (:import (java.io InputStream)))
+  (:import (java.io InputStream)
+           (org.eclipse.jetty.websocket.servlet ServletUpgradeResponse ServletUpgradeRequest)))
```
Comment (shamsimam) [inline waiter/src/waiter/handler.clj:52]: Lexicographic order

File: waiter/src/waiter/process_request.clj [hunk 13]
```diff
@@ -873,20 +875,41 @@
...
+  "Middleware to check for maintenance mode of a token and if improper maintenance mode header is provided"
+  [handler]
+  (make-maintenance-mode
+    handler
+    (fn [data-map request]
+      (-> data-map
+          (utils/data->error-response request)))
+    (fn [data-map request]
+      (-> data-map
+          (utils/data->maintenance-mode-response request)))))
```
Comment (shamsimam) [inline waiter/src/waiter/process_request.clj:902]: Nitpick: we can avoid using the arrow operator here as there is a single invocation.

File: waiter/src/waiter/process_request.clj [hunk 14]
```diff
@@ -873,20 +875,41 @@
...
             :else (handler request)))))
 
+(defn wrap-maintenance-mode
+  "Middleware to check for maintenance mode of a token and if improper maintenance mode header is provided"
+  [handler]
+  (make-maintenance-mode
+    handler
+    (fn [data-map request]
+      (-> data-map
+          (utils/data->error-response request)))
```
Comment (shamsimam) [inline waiter/src/waiter/process_request.clj:899]: Nitpick: we can avoid using the arrow operator here as there is a single invocation.

**Review by shamsimam — APPROVED** [review]

**Review by shamsimam — COMMENTED** [review]
Clicked approve by mistake, needs changes as per the previous feedback.

File: waiter/integration/waiter/websocket_integration_test.clj (hunk 6 above)
Comment (kevo1ution) [reply waiter/integration/waiter/websocket_integration_test.clj]: For some reason, the message that is sent using the ```ServletUpgradeResponse response``` object is not showing up in the error object here. That maintenance message gets lost somewhere in how the client handles the response

File: waiter/integration/waiter/websocket_integration_test.clj [hunk 15]
```diff
@@ -192,6 +216,139 @@
...
+          (let [token-response (post-token waiter-url token-description)]
+            (assert-response-status token-response http-200-ok)
+            (try
+              (let [response-promise (promise)
+                    connection (ws-client/connect!
+                                 (websocket-client-factory)
+                                 (ws-url waiter-url "/websocket-auth")
+                                 (fn [{:keys [in out]}]
+                                   (async/go
+                                     (async/>! out "request-info")
```
Comment (shamsimam) [inline waiter/integration/waiter/websocket_integration_test.clj:266]: We expect the upgrade request to fail. This block should 
- close `out` immediately
- write to an atom and to trigger an assertion failure later when we check the `assert-websocket-upgrade-error-response`.

File: waiter/integration/waiter/websocket_integration_test.clj [hunk 16]
```diff
@@ -192,6 +216,139 @@
...
```
Comment (shamsimam) [inline waiter/integration/waiter/websocket_integration_test.clj:243]: We expect the upgrade request to fail. This block should 
- close `out` immediately
- write to an atom and to trigger an assertion failure later when we check the `assert-websocket-upgrade-error-response`.

File: waiter/src/waiter/auth/authentication.clj [hunk 17]
```diff
@@ -379,16 +380,37 @@
...
+(defn- make-wrap-auth-bypass
+  "Takes a handler and a function to call when there is an error with processing the
+  authentication parameter. If successful and authentication is disabled for the token,
+  then the :skip-authentication is set to true for the request object before getting passed
+  to the handler"
+  [handler on-error]
   (fn auth-bypass-handler [{:keys [waiter-discovery] :as request}]
     (process-authentication-parameter
       waiter-discovery
       (fn [status message]
```
Comment (shamsimam) [inline waiter/src/waiter/auth/authentication.clj:392]: Nitpick: Can we add names for the functions defined via `fn`.

File: waiter/src/waiter/core.clj [hunk 18]
```diff
@@ -1881,4 +1863,11 @@
                                                  authentication-method-wrapper-fn)]
                                    (fn inner-wrap-secure-request-fn [{:keys [uri] :as request}]
                                      (log/debug "secure request received at" uri)
-                                     (handler request))))))})
+                                     (handler request))))))
+   :wrap-service-discovery-fn (pc/fnk [[:routines discover-service-parameters-fn]]
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj:1867]: Does this need to move from `routines`?

File: waiter/src/waiter/handler.clj [hunk 19]
```diff
@@ -48,23 +48,44 @@
...
-              (utils/attach-waiter-source)))
+          (on-redirect request))
         (handler request)))))
 
+(defn wrap-https-redirect
+  "Middleware that enforces https-redirect if a token has it configured before passing the request map to the next handler."
+  [handler]
+  (make-https-redirect
+    handler
+    #(-> (ssl/ssl-redirect-response % {})
```
Comment (shamsimam) [inline waiter/src/waiter/handler.clj:74]: Would prefer a `fn` with a name, e.g. `prepare-https-redirect-http-response`

File: waiter/src/waiter/handler.clj [hunk 20]
```diff
@@ -48,23 +48,44 @@
...
+         (utils/attach-waiter-source))))
+
+(defn wrap-wss-redirect
+  "Middleware that enforces https-redirect for websocket-request-acceptor if a token has it configured before passing
+  the request map to the next handler."
+  [handler]
+  (make-https-redirect
+    handler
+    (fn [{:keys [^ServletUpgradeResponse upgrade-response] :as request}]
```
Comment (shamsimam) [inline waiter/src/waiter/handler.clj:83]: Would prefer a `fn` with a name, e.g. `prepare-https-redirect-ws-upgrade-response`

File: waiter/src/waiter/main.clj [hunk 21]
```diff
@@ -125,7 +125,7 @@
                                                         core/correlation-id-middleware
                                                         (core/wrap-request-info router-id support-info)
                                                         consume-request-stream)
-                                        :websocket-acceptor websocket-request-acceptor
+                                        :websocket-acceptor (:websocket-request-acceptor handlers)
```
Comment (shamsimam) [inline waiter/src/waiter/main.clj:128]: Can we destructure this in the let block?

File: waiter/src/waiter/auth/authentication.clj [hunk 22]
```diff
@@ -379,16 +380,37 @@
...
+      (fn on-auth-enabled []
+        (handler request)))))
+
+(defn wrap-auth-bypass
+  "Middleware that checks if authentication is disabled for a token and sets the
+  :skip-authentication key of the request to true before passing to next handler."
+  [handler]
+  (make-wrap-auth-bypass
+    handler
+    (fn on-error [_ status message]
```
Comment (shamsimam) [inline waiter/src/waiter/auth/authentication.clj:405]: Each of these functions need unique names if they are in the same file :(
E.g. `on-http-error` and `on-ws-error`

File: waiter/src/waiter/main.clj [hunk 23]
```diff
@@ -107,11 +107,12 @@
    :scheduler core/scheduler
    :settings (pc/fnk dummy-symbol-for-fnk-schema-logic :- settings/settings-schema [] settings)
    :state core/state
-   :http-server (pc/fnk [[:routines discover-service-parameters-fn generate-log-url-fn waiter-request?-fn websocket-request-acceptor]
+   :http-server (pc/fnk [[:routines discover-service-parameters-fn generate-log-url-fn waiter-request?-fn]
                          [:settings cors-config host port server-options support-info websocket-config]
                          [:state cors-validator router-id server-name]
                          handlers] ; Insist that all systems are running before we start server
-                  (let [options (merge (cond-> server-options
+                  (let [websocket-request-acceptor (:websocket-request-acceptor handlers)
```
Comment (shamsimam) [inline waiter/src/waiter/main.clj:114]:
```suggestion
                  (let [{:keys [websocket-request-acceptor]{ handlers
```

File: waiter/src/waiter/process_request.clj [hunk 24]
```diff
@@ -873,20 +875,39 @@
...
+                                          :status http-503-service-unavailable}
+                                         request))
             :else (handler request)))))
 
+(defn wrap-maintenance-mode
+  "Middleware to check for maintenance mode of a token and if improper maintenance mode header is provided"
+  [handler]
+  (make-maintenance-mode
+    handler
+    (fn [data-map request]
```
Comment (shamsimam) [inline waiter/src/waiter/process_request.clj:897]: Nitpick: function names here and below

**Review by shamsimam — APPROVED** [review]
Awaiting green build

**Review by shamsimam — APPROVED** [review]
